### ISSUE TITLE
nonzero or zeroshot?

### ISSUE THREAD

**Issue by zhaoxinyu2002** [kg-cc/IA-GGAD#1]
Hello,

Thank you for your excellent NeurIPS Spotlight paper (IA-GGAD) and for making the code publicly available.

While reading the paper and attempting to reproduce the results, I noticed a critical discrepancy between the experimental setup described in the paper and the one implemented in the code. I am hoping you could provide some clarification.

**The paper's core claim is "Zero-shot".**

  * The paper repeatedly emphasizes that the model requires **"without any target-domain retraining or labeled data"** and operates **"without access to any target-domain labels or retraining"**.

**However, the provided code implementation appears to be "Few-shot" and relies on test set labels during the inference phase.**

My analysis of the code is as follows:

1.  In `run_IA_GGAD.py`, the script iterates through the **test dataset** (`data_test`) and calls `te_data.random_sample(args.shot)` for each test graph, with `--shot` defaulting to 10.

2.  In `utils.py`, the `random_sample` function accesses the **test graph's full labels** (`y = self.graph.ano_labels`). It then filters for **"normal" nodes** (`normal_idx = torch.where(y == 0)[0].tolist()`) and selects `shot` of them to create the `shot_mask`.

3.  In `train_test.py`, during the evaluation phase, the script passes both `test_graph.shot_mask` and the **full `test_graph.ano_labels`** to the scoring function `self.model.cross_attn.get_test_score`.

4.  Most importantly, in `model.py`, the `get_test_score` function uses these leaked labels to determine the support set for scoring:

    ```python
    negative_indices = torch.nonzero((prompt_mask == True) & (y == 0)).squeeze(1).tolist()
    ```

    This line seems to confirm that the model's inference is based on a support set of nodes that are **known to be "normal" (label 0)** from the test set.

**My Question:**

Is this the intended behavior for the code? If so, this seems to conflict with the "zero-shot" claim of the paper, as the inference process explicitly uses `shot` number of known normal labels from the target domain.

Am I misunderstanding how the code is intended to be run? Or is there perhaps another script or setting that allows for evaluation in a strict "zero-shot" setting (i.e., without accessing `y` in `get_test_score`)?

Thank you for your time and clarification\!

**Comment by kg-cc** [comment]
Thank you very much for your interest in our work and for your question regarding nonzero or zeroshot.
Since this work was compiled quite some time ago, the latest code had not yet been uploaded. The most recent version has now been uploaded.
Regarding your query: In the latest code,
the `utils.py` file uses `test_random_sample`, which actually yields `all_idx = list(range(num_nodes))`. This does not select normal nodes as few-shot examples.
The code is as follows:

    def test_random_sample(self, shot=10):
        y = self.graph.ano_labels
        num_nodes = y.shape[0]
        all_idx = list(range(num_nodes))
        random.shuffle(all_idx)
        shot_idx = torch.tensor(all_idx[:shot])
        shot_mask = torch.zeros(num_nodes, dtype=torch.bool)
        shot_mask[shot_idx] = True
        self.graph.shot_idx = shot_idx
        self.graph.shot_mask = shot_mask

The final `get_test_score` function is designed solely to retrieve the $RS(v_i)$ score during testing. It does not involve data leakage from specific normal or abnormal nodes, as the selection of `test_random_sample` is fully randomized.
` negative_indices = torch.nonzero((prompt_mask == True)).squeeze(1).tolist() `

I hope my response addresses your question.

**Comment by zhaoxinyu2002** [comment]
First, thank you for your work on the IA-GGAD paper and for open-sourcing your code.

While reviewing the code to understand your method, I found a significant discrepancy between the paper's "zero-shot" methodology and the practical implementation, which appears to be "few-shot" and relies on test-set labels.

I would be grateful if you could clarify this. Here is a summary of my findings:

-----

### 1\. Paper's "Zero-shot" Claim

The paper states it's a "zero-shot" framework and can make predictions **"without access to any target-domain labels"**.

-----

### 2\. Code Implementation Appears to be "Few-shot"

The `run_IA_GGAD.py` script calls `random_sample` for each **test graph**.

#### Evidence A: The `random_sample` function in `utils.py`

This function appears to access the test set labels (`y = self.graph.ano_labels`) and specifically select known "normal" nodes (`torch.where(y == 0)`).

```python
    def random_sample(self, shot=10):
        y = self.graph.ano_labels
        num_nodes = y.shape[0]
        normal_idx = torch.where(y == 0)[0].tolist()
        random.shuffle(normal_idx)
        shot_idx = torch.tensor(normal_idx[:shot])
        shot_mask = torch.zeros(num_nodes, dtype=torch.bool)
        self.graph.shot_idx = shot_idx
        shot_mask[shot_idx] = True
        self.graph.shot_mask = shot_mask
```

#### Evidence B: Similarity to ARC's `few_shot` function

I also noticed this `random_sample` function is functionally identical to the `few_shot` function from the ARC baseline (which your paper compares against).

*ARC's `few_shot`:*

```python
    def few_shot(self, shot=10):
        y = self.graph.ano_labels
        num_nodes = y.shape[0]
        normal_idx = torch.where(y == 0)[0].tolist()
        random.shuffle(normal_idx)
        shot_idx = torch.tensor(normal_idx[:shot])
        shot_mask = torch.zeros(num_nodes, dtype=torch.bool)
        self.graph.shot_idx = shot_idx
        shot_mask[shot_idx] = True
        self.graph.shot_mask = shot_mask
```

#### Evidence C: Use of Leaked Labels in `model.py`

The `get_test_score` function, which `train_test.py` calls for evaluation, appears to use these leaked labels. It explicitly filters for nodes that are **both** in the prompt set (`prompt_mask == True`) **and** are known normal (`y == 0`).

```python
    def get_test_score(self, X, codebook_sum, prompt_mask, y):
        # prompt node indices
        negative_indices = torch.nonzero((prompt_mask == True) & (y == 0)).squeeze(1).tolist()

        n_support_embed = X[negative_indices]

        # ... 

        query_tilde_embed = torch.mean(n_support_embed, dim=0).unsqueeze(0).repeat(len(query_indices), 1)
        # ...
        diff = query_embed - query_tilde_embed
        # ...
        query_score = torch.sqrt(torch.sum(diff ** 2, dim=1))
        # ...
        return test_score
```

-----

### My Question

Could you please clarify this? It appears the experimental results in Table 1 were generated using a "few-shot" setting, not a "zero-shot" one, by using this label-dependent `get_test_score` function.

This seems to fundamentally affect the paper's core "zero-shot" claim and the comparison to baselines.

Am I perhaps misunderstanding how the code is intended to be run, or was the wrong version of the code uploaded?

Thank you for your time and for helping me understand your work.

**Comment by kg-cc** [comment]
Thank you very much for your careful observation. I apologize for the confusion caused by the incorrect code version I uploaded.
The latest code has been uploaded.

    def test_random_sample(self, shot=10):
        y = self.graph.ano_labels
        num_nodes = y.shape[0]
        all_idx = list(range(num_nodes))
        random.shuffle(all_idx)
        shot_idx = torch.tensor(all_idx[:shot])
        shot_mask = torch.zeros(num_nodes, dtype=torch.bool)
        shot_mask[shot_idx] = True
        self.graph.shot_idx = shot_idx
        self.graph.shot_mask = shot_mask

    def get_test_score(self, X, codebook_sum, prompt_mask, y):
        # prompt node indices
        negative_indices = torch.nonzero((prompt_mask == True)).squeeze(1).tolist()
        n_support_embed = X[negative_indices]
        cos_sim_n = torch.matmul(n_support_embed, codebook_sum.T)
        top_idx_n = cos_sim_n.argmax(dim=1)
        code_n = codebook_sum[top_idx_n]
        query_indices = torch.nonzero(prompt_mask == False).squeeze(1).tolist()
        query_embed = X[query_indices]
        query_tilde_embed = torch.mean(n_support_embed, dim=0).unsqueeze(0).repeat(len(query_indices), 1)
        query_tilde_embed_code = torch.mean(code_n, dim=0).unsqueeze(0).repeat(len(query_indices), 1)
        diff = query_embed - query_tilde_embed
        diff_code = query_embed - query_tilde_embed_code

        query_score = torch.sqrt(torch.sum(diff ** 2, dim=1))
        query_score_code = torch.sqrt(torch.sum(diff_code ** 2, dim=1))
        test_score = (query_score + query_score_code) / 2
        return test_score
I hope the updated code resolves your issue. If you have any further questions, feel free to add me on WeChat: zx15877810381 for assistance.

**Comment by zhaoxinyu2002** [comment]
I understand now! Thank you for your patient explanations, and I look forward to the opportunity to collaborate!
